### ISSUE TITLE
PersistentCollection fails with on detached entities with FORCE_LAZY_LOADING

### ISSUE THREAD

**Issue by JWGmeligMeyling** [FasterXML/jackson-datatype-hibernate#98]
When `Feature.FORCE_LAZY_LOADING` is enabled, serialising a detached entity that has an non-initialised `PersistentCollection`, serialization fails with a `com.fasterxml.jackson.databind.JsonMappingException: failed to lazily initialize a collection of role: com.pallasathenagroup.entities.iris.Employee.employeeCapabilities, could not initialize proxy - no Session`.

I can somehow imaging why this is expected behaviour, but shouldn't lazy loading be treated differently for detached entities?

**Comment by walfrat** [comment]
This is because you're outside of the transaction. If you use spring to manage your transaction use the OpenSessionInViewFilter http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/orm/hibernate3/support/OpenSessionInViewFilter.html

**Comment by JWGmeligMeyling** [comment]
No its definetly not, because I just the exact same request filter as with any other lazy loading request. So It is within a transaction, with the entity manager open and just fine, only the entity being serialized is detached.

**Comment by chupetoide** [comment]
Hi Jan, maybe we are facing the same issue. I posted a [question on stackoverflow](http://stackoverflow.com/questions/40773005/how-to-serialize-and-deserialize-a-set-attribute-with-jackson-and-hibernate)  but nobody has answered yet. I am trying to export database objects to JSON using JPA+Hibernate+Jackson. If a stored object has a List or Set attribute, it is serialized as PersistentBag or PersistentSet (hibernate objects). So when I try to deserialize the JSON it tries to load a PersistentBag and throws the same exception you have.

**Comment by JWGmeligMeyling** [comment]
The whole idea of this Hibernate integration is to be able to serialize persistent sets and bags and Hibernate proxies with Jackson ;) So normally this just works. That is except with both this specific feature enabled AND trying to serialize a detached entity (see entitymanager.detach documentation).

**Comment by chupetoide** [comment]
If I detach the model before serialization it throws the same exception when trying to serialize the Set attribute:

```java
Hibernate4Module h4m = new Hibernate4Module();
h4m.enable(Hibernate4Module.Feature.FORCE_LAZY_LOADING);
		
ObjectMapper mapper = new ObjectMapper();
mapper.registerModule(h4m);
mapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
mapper.enableDefaultTyping();
		
TestCar storedCar = PersistenceManager.get(TestCar.class, 1L);
PersistenceManager.getEntityManager().detach(storedCar);
String jsonContent = mapper.writeValueAsString(storedCar);
```

At this point I am not sure if the problem is on the serialization process (translating Set to PersistentSet) or on the deserialization process (failing to load the Set from json).

**Comment by JWGmeligMeyling** [comment]
Yes, you can't serialise an persistent collection from a detached entity, unless its loaded in advance. That actually makes sense, because you need a session in order to be able to fill the collection, and you just specifically asked to disconnect the entity from the session. That part makes sense, the exception could be a bit more descriptive though, and it's debatable what actually is intended behaviour for `Hibernate4Module.Feature.FORCE_LAZY_LOADING` on detached entities (for example failing back to serializing a `null`, as it would do without `Hibernate4Module.Feature.FORCE_LAZY_LOADING` on enabled). Anyway, if you need to detach your entity before serialization , you can use `Hibernate.initialize(storedCar.getMyCollection()`. Or you simply don't detach the entity, which is what I usually go for :wink:

**Comment by chupetoide** [comment]
Let me explain it a bit more. I have no issues on the serialization process. It works fine and all data is serialized as well. My problem is when I try to deserialize the generated JSON because it fails instantiating the PersistentSet. This is the generated JSON I am trying to deserialize:

```json
{
  "id" : 1,
  "name" : "Fiat",
  "parts" : [ "org.hibernate.collection.internal.PersistentSet", [ [ "com.acme.TestPart", {
    "id" : 1,
    "name" : "puerta"
  } ] ] ]
}
```

**Comment by suikast42** [comment]
I know  it's a little bit old topic but i am facing on the the promlem with antoher serilisation framework names kryo. 

> @chupetoide  Let me explain it a bit more. I have no issues on the serialization process. It works fine and all data is serialized as well. My problem is when I try to deserialize the generated JSON because it fails instantiating the PersistentSet. This is the generated JSON I am trying to deserialize:

Exactly that is my situation too.

But that is a hibernate problem. Hibernate geneartes for managed relations an instance of PersistentCollection which needs a hibernate session for collection operations. 

But have not a soluition for that. Maybe a DTO pattern is the only soliton for that?

**Comment by ashraf-revo** [comment]
any help about this bug, or how to map types when deserialize from `org.hibernate.collection.internal.PersistentBag` to `java.util.ArrayList`

**Comment by jwgmeligmeyling** [comment]
Its hardly a bug rather expected behaviour, and workarounds have already been suggested above.

An enhancement for this issue would obviously be nice to have but serializing detached entity completely conflicts with the idea of forcing lazy initialization during serialization anyway.

**Comment by ashraf-revo** [comment]
it work with me when i used Hibernate5Module with enabling `REPLACE_PERSISTENT_COLLECTIONS,FORCE_LAZY_LOADING`![alt text](https://i.ibb.co/gJZ7Qh1/Screenshot-from-2021-05-20-01-55-22.png)

**Comment by sharkdeng** [comment]
I can load class Hibernate5Module
